### PR TITLE
Storage Container/Share retry create if 409 && beingdeleted

### LINKED ISSUE #6803: Can't delete and immediately recreate a storage container
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.12.24
"azurerm" (hashicorp/azurerm) 2.2.0

### Affected Resource(s)

* `azurerm_storage_container`

### Terraform Configuration Files

```hcl
resource "azurerm_storage_container" "capture_storage_container" {
  name                  = lower(var.area_name)
  storage_account_name  = var.capture_storage_account_name
  container_access_type = "private"
}
```

### Debug Output

```
terraform destroy
azurerm_storage_container.capture_storage_container: Destroying... [id=https://strgappmercurycd.blob.core.windows.net/master]
azurerm_storage_container.capture_storage_container: Destruction complete after 0s
```

```
terraform apply
Acquiring state lock. This may take a few moments...
azurerm_storage_container.capture_storage_container: Creating...
Error: Error creating Container "master" (Account "strgappmercurycd" / Resource Group “ci-rg-001"): containers.Client#Create: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="ContainerBeingDeleted" Message="The specified container is being deleted. Try operation later.\nRequestId:38c204cc-f01e-001f-7cc4-239161000000\nTime:2020-05-06T16:36:08.1842227Z"

  on main.tf line 15, in resource "azurerm_storage_container" "capture_storage_container":
  15: resource "azurerm_storage_container" "capture_storage_container" {
```
```


### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

<!--- What should have happened? --->
For integration testing we delete and reprovision the full environment immediately. Thus it should be possible to delete a storage container and recreate it immediately afterwards.

### Actual Behavior

<!--- What actually happened? --->
Terraform apply failed.

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform destroy`
1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

* #0000

### PR DESCRIPTION
**Pull request by nickmhankins**
Fixes #6803

### CODE REVIEW

**Comment by jackofallops** [comment]
Hi @nickmhankins 

Thanks for the effort on identifying this and the work on addressing it. 

I hope you don't mind, but I changed the code to a poller pattern we use to check state, rather than a flat 300s pause and added some tests that exhibit the issue identified to check the code works correctly.  They're passing locally for me, so I'm just running them on our CI system.  As soon as they're green and Travis is happy I'll get this merged.

Thanks again!

File: azurerm/internal/services/storage/resource_arm_storage_share.go [hunk 1]
```diff
@@ -166,8 +169,25 @@ func resourceArmStorageShareCreate(d *schema.ResourceData, meta interface{}) err
...
-		return fmt.Errorf("Error creating Share %q (Account %q / Resource Group %q): %+v", shareName, accountName, account.ResourceGroup, err)
+
+	if resp, err := client.Create(ctx, accountName, shareName, input); err != nil {
+		// If we fail due to previous delete still in progress, then we can retry
+		if utils.ResponseWasConflict(resp) && strings.Contains(err.Error(), "ShareBeingDeleted") {
+			stateConf := &resource.StateChangeConf{
+				Pending:        []string{"waitingOnDelete"},
+				Target:         []string{"succeeded"},
+				Refresh:        storageShareCreateRefreshFunc(ctx, client, accountName, shareName, input),
+				PollInterval:   30 * time.Second,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/storage/resource_arm_storage_share.go:180]: this is a data plane API, so let's reduce this down to 10-15s?

File: azurerm/internal/services/storage/resource_arm_storage_container.go [hunk 2]
```diff
@@ -128,8 +131,24 @@ func resourceArmStorageContainerCreate(d *schema.ResourceData, meta interface{})
...
-	if _, err := client.Create(ctx, accountName, containerName, input); err != nil {
-		return fmt.Errorf("Error creating Container %q (Account %q / Resource Group %q): %s", containerName, accountName, account.ResourceGroup, err)
+	if resp, err := client.Create(ctx, accountName, containerName, input); err != nil {
+		// If we fail due to previous delete still in progress, then we can retry
+		if utils.ResponseWasConflict(resp.Response) && strings.Contains(err.Error(), "ContainerBeingDeleted") {
+			stateConf := &resource.StateChangeConf{
+				Pending:        []string{"waitingOnDelete"},
+				Target:         []string{"succeeded"},
+				Refresh:        storageContainerCreateRefreshFunc(ctx, client, accountName, containerName, input),
+				PollInterval:   30 * time.Second,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/storage/resource_arm_storage_container.go:141]: this is a data plane API, so let's reduce this down to 10-15s?

**Comment by jackofallops** [comment]
Tests passing:
image

**Pull request merged**